datatype/list.go

```diff
@@ -131,12 +130,39 @@ func (c *ListCodec) Encode(data interface{}, version primitive.ProtocolVersion)
 	return buf.Bytes(), nil
 }
 
-func (c *ListCodec) Decode(encoded []byte, _ primitive.ProtocolVersion) (value interface{}, err error) {
-	// TODO
-	return nil, nil
+func (c *ListCodec) Decode(encoded []byte, version primitive.ProtocolVersion) (value interface{}, err error) {
+	if encoded == nil {
+		return nil, nil
+	}
+
+	n, read, err := readCollectionSize(version, encoded)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(encoded) < n {
+		return nil, fmt.Errorf("decode list: unexpected eof")
+	}
+
+	encoded = encoded[read:]
+	childDecodeOutputType := c.ElementCodec.GetDecodeOutputType()
```

Comment:
I'm curious, why did you have to introduce reflection and this new `GetDecodedOutputType` method? I'm a bit wary that reflection might introduce a performance penalty.

For example, what I had in mind for `ListCodec.Decode` would rather look like this:

```
func (c *ListCodec) Decode(encoded []byte, version primitive.ProtocolVersion) (value interface{}, err error) {
        // first lines omitted
	encoded = encoded[read:]
	slice := make([]interface{}, n, n)
	for i := 0; i < n; i++ {
		decoded, m, err := decodeChildElement(c.ElementCodec, encoded, version)
		if err != nil {
			return nil, err
		}
		encoded = encoded[m:]
		slice[i]= decoded
	}
	return slice, nil
}

func decodeChildElement(
	elementCodec Codec,
	encoded []byte,
	version primitive.ProtocolVersion) (output interface{}, read int, err error) {
        // first lines omitted
	var decoded interface{}
	if m < 0 {
		decoded = nil
	} else if len(encoded) < m {
		return nil, -1, fmt.Errorf("decode list: unexpected eof")
	} else {
		decoded, err = elementCodec.Decode(encoded[:m], version)
		if err != nil {
			return nil, -1, err
		}
	}
	return decoded, read + m, nil
}
```

Reply:
Ideally, we wouldn't need the `decodeChildElement` function at all.

This is more what I had in mind for the general logic of a collection codec:

```
encoded = encoded[read:]
slice := make([]interface{}, n, n)
for i := 0; i < n; i++ {
	decoded, err := c.ElementCodec.Decode(encoded, version)
	if err != nil {
		return nil, err
	}
	encoded = encoded[m:]
	slice[i]= decoded
}
return slice, nil
```

But this doesn't compile :-( We can't update the slice since we don't know how much was read from it.

That's why I was advocating for `io.Reader` instead.

Reply:
I used reflection because otherwise it is a PITA for the user to convert slices/maps of `interface{}` to the expected type. I assumed that this would have a performance hit but since we aren't planning to use these codecs for any performance sensitive tasks, I figured I shouldn't spent too much time optimizing something that we don't need atm.

Reply:
One way to avoid using reflection would be to receive a pointer as an argument similar to what gocql does. The caller would be responsible for providing a pointer to a type that is supported according to the CQL type that is being decoded. This approach doesn't seem too user friendly though.